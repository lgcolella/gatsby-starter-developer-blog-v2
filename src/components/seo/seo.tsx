/* Vendor imports */
import React from 'react'
import PropTypes from 'prop-types'
import { Helmet } from 'react-helmet'
import { StaticQuery, graphql } from 'gatsby'
/* App imports */
import { ChildImageSharpFixed } from '../../types'
import Config from '../../../config'
import Utils from '../../utils'

export const PureSEO = ({
  queryData,
  title,
  description,
  path,
  lang,
  contentType,
  imageUrl,
  keywords,
  translations,
}: PropsForPure) => {
  let pageUrl = Utils.resolveUrl(Config.siteUrl, Config.pathPrefix, path)
  let metaImageUrl = Utils.resolveUrl(
    Config.siteUrl,
    imageUrl ? imageUrl : queryData.file.childImageSharp.fixed.src
  )

  return (
    <Helmet
      title={title}
      titleTemplate={`%s | ${Config.siteTitle}`}
      meta={[
        { name: 'description', content: description }, // Page description
        /* Open Graph */
        { property: 'og:title', content: title },
        { property: 'og:type', content: contentType || 'website' },
        { property: 'og:url', content: pageUrl },
        { property: 'og:description', content: description },
        { property: 'og:image', content: metaImageUrl },
        { property: 'og:image:alt', content: description },
        { property: 'og:site_name', content: Config.siteTitle },
        { property: 'og:locale', content: lang || 'en_US' },
        /* Twitter card */
        { name: 'twitter:card', content: 'summary_large_image' },
        { name: 'twitter:title', content: title },
        { name: 'twitter:description', content: description },
        { name: 'twitter:image', content: metaImageUrl },
        { name: 'twitter:image:alt', content: description },
        { name: 'twitter:site', content: Config.author },
        { name: 'twitter:creator', content: Config.author },
      ].concat(
        keywords && keywords.length > 0
          ? { name: 'keywords', content: keywords.join(', ') }
          : []
      )}
      link={[{ rel: 'canonical', href: pageUrl }] // Canonical url
        // Translated versions of page
        .concat(
          translations
            ? translations.map(obj => ({
                rel: 'alternate',
                hreflang: obj.hreflang,
                href: Utils.resolveUrl(
                  Config.siteUrl,
                  Config.pathPrefix,
                  obj.path
                ),
              }))
            : []
        )}
    />
  )
}

PureSEO.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  path: PropTypes.string.isRequired,
  lang: PropTypes.string,
  contentType: PropTypes.oneOf(['article', 'website']),
  imageUrl: PropTypes.string,
  keywords: PropTypes.arrayOf(PropTypes.string),
  translations: PropTypes.arrayOf(
    PropTypes.shape({
      hreflang: PropTypes.string.isRequired,
      path: PropTypes.string.isRequired,
    })
  ),
  queryData: PropTypes.shape({
    file: PropTypes.shape({
      childImageSharp: PropTypes.shape({
        fixed: PropTypes.shape({
          src: PropTypes.string.isRequired,
        }).isRequired,
      }).isRequired,
    }).isRequired,
  }).isRequired,
}

PureSEO.defaultProps = {
  keywords: Config.siteKeywords,
}

export const SEO = (props: Props) => (
  <StaticQuery
    query={graphql`
      query {
        file(name: { eq: "facebook-logo" }) {
          childImageSharp {
            fixed(width: 500) {
              ...GatsbyImageSharpFixed_noBase64
            }
          }
        }
      }
    `}
    render={queryData => <PureSEO queryData={queryData} {...props} />}
  />
)

interface Props {
  title: string
  description: string
  path: string
  lang?: string
  contentType?: 'article' | 'website'
  imageUrl?: string
  keywords?: string[]
  translations?: Array<{
    hreflang: string
    path: string
  }>
}

interface PropsForPure extends Props {
  queryData: {
    file: {
      childImageSharp: ChildImageSharpFixed
    }
  }
}

export default SEO
