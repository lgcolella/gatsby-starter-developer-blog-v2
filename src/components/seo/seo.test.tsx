/* Vendor imports */
import React from 'react'
import { shallow } from 'enzyme'
/* App imports */
import { PureSEO } from './seo'
import { ImageSharpFixed } from '../../../tests/mocks'

describe('SEO component', () => {
  test('matches snapshot with minimal props', () => {
    let wrapper = shallow(
      <PureSEO
        title={'Title of article'}
        description={'Description of article'}
        path={'/path-of-article'}
        queryData={{
          file: {
            childImageSharp: ImageSharpFixed,
          },
        }}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  test('matches snapshot with all props', () => {
    let wrapper = shallow(
      <PureSEO
        title={'Title of article'}
        description={'Description of article'}
        path={'/path-of-article'}
        lang={'it'}
        contentType={'article'}
        imageUrl={'image-of-article.jpg'}
        keywords={['tag1', 'tag2']}
        translations={[
          { hreflang: 'it', path: 'it/path-of-article' },
          { hreflang: 'en', path: 'en/path-of-article' },
        ]}
        queryData={{
          file: {
            childImageSharp: ImageSharpFixed,
          },
        }}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
