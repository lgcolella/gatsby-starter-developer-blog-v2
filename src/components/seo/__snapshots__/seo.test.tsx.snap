// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO component matches snapshot with all props 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  link={
    Array [
      Object {
        "href": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/path-of-article/",
        "rel": "canonical",
      },
      Object {
        "href": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/it/path-of-article/",
        "hreflang": "it",
        "rel": "alternate",
      },
      Object {
        "href": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/en/path-of-article/",
        "hreflang": "en",
        "rel": "alternate",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Description of article",
        "name": "description",
      },
      Object {
        "content": "Title of article",
        "property": "og:title",
      },
      Object {
        "content": "article",
        "property": "og:type",
      },
      Object {
        "content": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/path-of-article/",
        "property": "og:url",
      },
      Object {
        "content": "Description of article",
        "property": "og:description",
      },
      Object {
        "content": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/image-of-article.jpg",
        "property": "og:image",
      },
      Object {
        "content": "Description of article",
        "property": "og:image:alt",
      },
      Object {
        "content": "Vincent van Gogh",
        "property": "og:site_name",
      },
      Object {
        "content": "it",
        "property": "og:locale",
      },
      Object {
        "content": "summary_large_image",
        "name": "twitter:card",
      },
      Object {
        "content": "Title of article",
        "name": "twitter:title",
      },
      Object {
        "content": "Description of article",
        "name": "twitter:description",
      },
      Object {
        "content": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/image-of-article.jpg",
        "name": "twitter:image",
      },
      Object {
        "content": "Description of article",
        "name": "twitter:image:alt",
      },
      Object {
        "content": "Vincent van Gogh",
        "name": "twitter:site",
      },
      Object {
        "content": "Vincent van Gogh",
        "name": "twitter:creator",
      },
      Object {
        "content": "tag1, tag2",
        "name": "keywords",
      },
    ]
  }
  title="Title of article"
  titleTemplate="%s | Vincent van Gogh"
/>
`;

exports[`SEO component matches snapshot with minimal props 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  link={
    Array [
      Object {
        "href": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/path-of-article/",
        "rel": "canonical",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Description of article",
        "name": "description",
      },
      Object {
        "content": "Title of article",
        "property": "og:title",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/path-of-article/",
        "property": "og:url",
      },
      Object {
        "content": "Description of article",
        "property": "og:description",
      },
      Object {
        "content": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/image-for-meta.png",
        "property": "og:image",
      },
      Object {
        "content": "Description of article",
        "property": "og:image:alt",
      },
      Object {
        "content": "Vincent van Gogh",
        "property": "og:site_name",
      },
      Object {
        "content": "en_US",
        "property": "og:locale",
      },
      Object {
        "content": "summary_large_image",
        "name": "twitter:card",
      },
      Object {
        "content": "Title of article",
        "name": "twitter:title",
      },
      Object {
        "content": "Description of article",
        "name": "twitter:description",
      },
      Object {
        "content": "https://lgcolella.github.io/gatsby-starter-developer-blog-v2/image-for-meta.png",
        "name": "twitter:image",
      },
      Object {
        "content": "Description of article",
        "name": "twitter:image:alt",
      },
      Object {
        "content": "Vincent van Gogh",
        "name": "twitter:site",
      },
      Object {
        "content": "Vincent van Gogh",
        "name": "twitter:creator",
      },
      Object {
        "content": "developer, development, web, javascript, frontend, backend, fullstack",
        "name": "keywords",
      },
    ]
  }
  title="Title of article"
  titleTemplate="%s | Vincent van Gogh"
/>
`;
